### ISSUE TITLE
empty to restrictive

### ISSUE THREAD

**Issue by DarkWingMcQuack** [ericniebler/range-v3#834]
I see myself often using the following to check if a range is empty:
```
template<class Rng>
auto is_empty(Rng&& r)
{
    return any_of(r,
                  [](auto&&) {
                      return true;
                  });
}
```
which works for all ranges and is in O(1). So why not using this to make empty work for all ranges.
If this is a stupid thing to do please let me know.

**Comment by ericniebler** [comment]
If your range is an Input range, like an istream, you have just consumed the first element of your range merely by asking whether the range is empty. That is why `ranges::empty` has the restrictions it does.

**Comment by CaseyCarter** [comment]
Expanding on this a bit: `ranges::empty` used to permit input ranges, and it was a bug farm. People don't realize that `begin` is not idempotent for input ranges: it's UB to call it repeatedly. When we hide the `begin` call inside of `ranges::empty`, it's easy to forget even to those that *know* repeated calls to begin are UB.

As a data point, we decided to make `ranges::empty` require forward ranges after I implemented the especially nasty `debug_input_view` which asserts on violations of the minimal input range requirements - including repeated calls to `begin`. I found maybe half a dozen occurrences in range-v3 itself where we were passing input ranges to `empty` or `front` and later calling `begin` a second time.
